pkg/frontend/v2/frontend.go

```diff
@@ -238,11 +245,22 @@ func (f *Frontend) RoundTripGRPC(ctx context.Context, req *httpgrpc.HTTPRequest)
 		// Buffer of 1 to ensure response or error can be written to the channel
 		// even if this goroutine goes away due to client context cancellation.
 		enqueue:  make(chan enqueueResult, 1),
-		response: make(chan *frontendv2pb.QueryResultRequest, 1),
+		response: make(chan queryResultWithBody, 1),
 	}
 
 	f.requests.put(freq)
-	defer f.requests.delete(freq.queryID)
+	// delete is called through the cleanup func executed either in the defer or by the caller closing the body.
+
+	cleanup := func() {
+		defer cancel(errExecutingQueryRoundTripFinished)
+		f.requests.delete(freq.queryID)
```

Comment:
I used `defer` for both calls now to keep the order of execution the same as it was. If you prefer we can also remove the defer and flip the order, wdyt?

Reply:
i just didn't see the use for `defer` in such a small function. Technically using a defer means it will run even after a panic, but I don't think that was part of your intention, was it?

Reply:
My intention was to make these cleanup actions run under the same conditions / in the same order as they were before this change. Should we make this

```suggestion
		f.requests.delete(freq.queryID)
		cancel(errExecutingQueryRoundTripFinished) 
```
(with flipped order of the statements compared to before) instead?

Reply:
makes sense. I meant the same in my original comment

Reply:
Changed in 3b315fb0.